src/sources/exec.rs

```diff
@@ -0,0 +1,843 @@
+        stdout_reader,
+        shutdown.clone(),
+        config.event_per_line,
+        config.maximum_buffer_size,
+        STDOUT,
+        sender,
+    );
+
+    while let Some((line, stream)) = receiver.recv().await {
+        let event = create_event(
+            &config,
+            Bytes::from(line),
+            &Some(stream.to_string()),
+            Some(pid),
+            None,
+            &None,
+        );
+
+        let _ = out
+            .send(event)
+            .await
+            .map_err(|_: crate::pipeline::ClosedError| {
+                error!(message = "Failed to forward events; downstream is closed.");
+            });
+    }
+
+    let elapsed = now.elapsed();
+
+    info!("Finished command run.");
+    let _ = out.flush().await;
+
+    // TODO: Tokio 1.0.1+ has a wait and try_wait method to get exit status
+    if let Poll::Ready(Ok(exit_status)) = futures::poll!(child) {
+        handle_exit_status(&config, pid, exit_status, elapsed, out).await;
+        Ok(Some(exit_status))
+    } else {
+        emit!(ExecCommandExecuted {
+            command: config.command.as_str(),
+            exit_status: None,
+            exec_duration: elapsed,
+        });
+
+        Ok(None)
+    }
+}
+
+async fn handle_exit_status(
+    config: &ExecConfig,
+    pid: u32,
+    exit_status: ExitStatus,
+    exec_duration: Duration,
+    mut out: Pipeline,
+) {
+    emit!(ExecCommandExecuted {
+        command: config.command.as_str(),
+        exit_status: exit_status.code(),
+        exec_duration,
+    });
+
+    let event = create_event(
+        config,
+        Bytes::new(),
+        &None,
+        Some(pid),
+        exit_status.code(),
+        &Some(exec_duration.as_millis()),
+    );
+
+    let _ = out
+        .send(event)
+        .await
+        .map_err(|_: crate::pipeline::ClosedError| {
+            error!(message = "Failed to forward events; downstream is closed.");
+        });
+}
+
+fn build_command(config: &ExecConfig) -> Command {
+    let mut command = Command::new(config.command.as_str());
+    command.kill_on_drop(true);
+
+    // Explicitly set the current dir if needed
+    if let Some(current_dir) = &config.current_dir {
+        command.current_dir(current_dir);
+    }
+
+    // Pipe our stdout/stderr to the process to we inherit it's output
+    command.stdout(std::process::Stdio::piped());
+    command.stderr(std::process::Stdio::piped());
+
+    if let Some(arguments) = &config.arguments {
+        if !arguments.is_empty() {
+            command.args(arguments);
+        }
+    }
+
+    command
+}
+
+fn create_event(
+    config: &ExecConfig,
+    line: Bytes,
+    data_stream: &Option<String>,
+    pid: Option<u32>,
+    exit_status: Option<i32>,
+    exec_duration_millis: &Option<u128>,
+) -> Event {
+    emit!(ExecEventReceived {
+        command: config.command.as_str(),
+        byte_size: line.len(),
+    });
+
+    let mut event = Event::from(line);
+    let log_event = event.as_mut_log();
```

Comment:
This is obsolete usage for new code. Use `let mut log_event = LogEvent::default();` and then return `log_event.into()` or `Event::Log(log_event)`

Reply:
Fixed. I explicitly added `message` and `timestamp`.